src/compiler/transformers/automatic-key-insertion/index.ts

```diff
@@ -0,0 +1,136 @@
+import ts from 'typescript';
+
+import { deriveJSXKey } from './utils';
+
+/**
+ * A transformer factory to create a transformer which will add `key`
+ * properties to all of the JSX nodes contained inside of a component's `render`
+ * function.
+ *
+ * This can be thought of as transforming the following:
+ *
+ * ```tsx
+ * class MyComponent {
+ *   render() {
+ *     <div>hey!</div>
+ *   }
+ * }
+ * ```
+ *
+ * to this:
+ *
+ * ```tsx
+ * class MyComponent {
+ *   render() {
+ *     <div key="a-unique-key">hey!</div>
+ *   }
+ * }
+ * ```
+ *
+ * The inserted keys are generated by {@link deriveJSXKey}.
+ *
+ * @param transformCtx a transformation context
+ * @returns a typescript transformer for inserting keys into JSX nodes
+ */
+export const performAutomaticKeyInsertion = (transformCtx: ts.TransformationContext): ts.Transformer<ts.SourceFile> => {
+  /**
+   * This is our outer-most visitor function which serves to locate a class
+   * declaration, at which point it hands things over to the next visitor
+   * function ({@link findRenderMethodVisitor}) which locates the `render`
+   * method.
+   *
+   * @param node a typescript syntax tree node
+   * @returns the result of handling the node
+   */
+  function findClassDeclVisitor(node: ts.Node): ts.VisitResult<ts.Node> {
+    if (ts.isClassDeclaration(node)) {
+      return ts.visitEachChild(node, findRenderMethodVisitor, transformCtx);
+    } else {
+      return ts.visitEachChild(node, findClassDeclVisitor, transformCtx);
+    }
+  }
+
+  /**
+   * This middle visitor function is responsible for finding the render method
+   * on a Stencil class and then passing off responsibility to the inner-most
+   * visitor, which deals with syntax nodes inside the method.
+   *
+   * @param node a typescript syntax tree node
+   * @returns the result of handling the node
+   */
+  function findRenderMethodVisitor(node: ts.Node): ts.VisitResult<ts.Node> {
+    if (ts.isMethodDeclaration(node) && node.name.getText() === 'render') {
+      return ts.visitEachChild(node, jsxElementVisitor, transformCtx);
```

Comment:
mm that's a good point, we probably would want to do something like that anyway to avoid transforming someone's weird in-a-stencil-project React components or something

I'll pull those lines form `convert-decorators` out into a helper (`classDeclIsStencilComponent` or something)

Reply:
So! I wasn't able to use the code from `convert-decorators` here because the relevant syntax nodes have already been removed from the component class node when they're 'static-ized' in `convertDecoratorsToStatic`, _however_ I was able to add a check which looks for the `is` property, as we already do here in `parseStaticComponentMeta`:

https://github.com/ionic-team/stencil/blob/cea23c4a306d05a41f9ca137e9168039276ba30a/src/compiler/transformers/static-to-meta/component.ts#L49-L53

this lets me determine whether or not a class is a Stencil component, and gate the transformation on that. I added a test to ensure that a non-decorated class which has a `render()` method with JSX doesn't get transformed.